--- bfd/bfd.c	2015-11-13 09:27:40.000000000 +0100
+++ bfd/bfd.c	2016-01-29 19:12:23.918144600 +0100
@@ -1977,9 +1977,8 @@
 {
   if ((abfd->flags & BFD_COMPRESS) != 0)
     {
-      if (bfd_get_flavour (abfd) == bfd_target_elf_flavour)
-	{
-	  if ((abfd->flags & BFD_COMPRESS_GABI) != 0)
+      if (bfd_get_flavour (abfd) == bfd_target_elf_flavour
+	  && (abfd->flags & BFD_COMPRESS_GABI) != 0)
 	    {
 	      const struct elf_backend_data *bed
 		= get_elf_backend_data (abfd);
@@ -2010,7 +2009,8 @@
 	  else
 	    {
 	      /* Clear the SHF_COMPRESSED bit.  */
-	      elf_section_flags (sec) &= ~SHF_COMPRESSED;
+	      if (bfd_get_flavour (abfd) == bfd_target_elf_flavour)
+		elf_section_flags (sec) &= ~SHF_COMPRESSED;
 
 	      /* Write the zlib header.  It should be "ZLIB" followed by
 		 the uncompressed section size, 8 bytes in big-endian
@@ -2018,7 +2018,6 @@
 	      memcpy (contents, "ZLIB", 4);
 	      bfd_putb64 (sec->size, contents + 4);
 	    }
-	}
     }
   else
     abort ();
--- gas/write.c	2015-11-13 09:27:41.000000000 +0100
+++ gas/write.c	2016-01-29 17:56:26.360467200 +0100
@@ -1428,7 +1428,8 @@
   if (strm == NULL)
     return;
 
-  if (flag_compress_debug == COMPRESS_DEBUG_GABI_ZLIB)
+  if (flag_compress_debug == COMPRESS_DEBUG_GABI_ZLIB
+      && bfd_get_flavour (abfd) == bfd_target_elf_flavour)
     {
       stdoutput->flags |= BFD_COMPRESS | BFD_COMPRESS_GABI;
       compression_header_size
