From 3bcfef7747331a71953919a8cf7448342d089600 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sun, 28 Mar 2021 14:00:25 +0200
Subject: [PATCH 01/36] Fix gengtype for windows paths

---
 gcc/gengtype.cc | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/gcc/gengtype.cc b/gcc/gengtype.cc
index c0ecbe8558b..137c4f90bd6 100644
--- a/gcc/gengtype.cc
+++ b/gcc/gengtype.cc
@@ -5081,7 +5081,10 @@ parse_program_options (int argc, char **argv)
 	  break;
 	case 'S':		/* --srcdir */
 	  if (optarg)
+	  {
 	    srcdir = optarg;
+	    if( srcdir[1]==':' ) srcdir += 2;
+	  }
 	  else
 	    fatal ("missing source directory");
 	  srcdir_len = strlen (srcdir);
-- 
2.39.1.windows.1

