From 4c814301e101653eff9c45f4c419fe67d4814284 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sun, 28 Mar 2021 14:00:25 +0200
Subject: [PATCH 01/40] Fix gengtype for windows paths

---
 gcc/gengtype.cc | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/gcc/gengtype.cc b/gcc/gengtype.cc
index 6a3621c96cd..872db894cf6 100644
--- a/gcc/gengtype.cc
+++ b/gcc/gengtype.cc
@@ -5087,7 +5087,10 @@ parse_program_options (int argc, char **argv)
 	  break;
 	case 'S':		/* --srcdir */
 	  if (optarg)
+	  {
 	    srcdir = optarg;
+	    if( srcdir[1]==':' ) srcdir += 2;
+	  }
 	  else
 	    fatal ("missing source directory");
 	  srcdir_len = strlen (srcdir);
-- 
2.54.0

