From 2920d370eb1956f0845d6e5a485f36098f0ff770 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sun, 28 Mar 2021 14:00:25 +0200
Subject: [PATCH 01/39] Fix gengtype for windows paths

---
 gcc/gengtype.cc | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/gcc/gengtype.cc b/gcc/gengtype.cc
index 5860f9875ed..7fd7373554e 100644
--- a/gcc/gengtype.cc
+++ b/gcc/gengtype.cc
@@ -5088,7 +5088,10 @@ parse_program_options (int argc, char **argv)
 	  break;
 	case 'S':		/* --srcdir */
 	  if (optarg)
+	  {
 	    srcdir = optarg;
+	    if( srcdir[1]==':' ) srcdir += 2;
+	  }
 	  else
 	    fatal ("missing source directory");
 	  srcdir_len = strlen (srcdir);
-- 
2.53.0

