From df5106e1488ee9a16fc693999958b50d5aec0b2e Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sun, 28 Mar 2021 14:00:25 +0200
Subject: [PATCH 01/15] Fix gengtype for windows paths

---
 gcc/gengtype.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/gcc/gengtype.c b/gcc/gengtype.c
index 98d4626f87e..5c74fdf4cd2 100644
--- a/gcc/gengtype.c
+++ b/gcc/gengtype.c
@@ -5038,7 +5038,10 @@ parse_program_options (int argc, char **argv)
 	  break;
 	case 'S':		/* --srcdir */
 	  if (optarg)
+	  {
 	    srcdir = optarg;
+	    if( srcdir[1]==':' ) srcdir += 2;
+	  }
 	  else
 	    fatal ("missing source directory");
 	  srcdir_len = strlen (srcdir);
-- 
2.15.1.windows.2

