From 278967352599b91dc92363257042b7eb23c734f3 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sun, 28 Mar 2021 14:34:17 +0200
Subject: [PATCH 07/17] Disable #ident directive by default

---
 gcc/common.opt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gcc/common.opt b/gcc/common.opt
index 8a0dafc522d..ce0cfd82a93 100644
--- a/gcc/common.opt
+++ b/gcc/common.opt
@@ -1775,7 +1775,7 @@ Harden conditional branches by checking reversed conditions.
 ; On SVR4 targets, it also controls whether or not to emit a
 ; string identifying the compiler.
 fident
-Common Var(flag_no_ident,0)
+Common Var(flag_no_ident,0) Init(1)
 Process #ident directives.
 
 fif-conversion
-- 
2.39.1.windows.1

