From 0360d8ed3d007a2737eb04454a117cdf70e913ad Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sun, 28 Mar 2021 14:35:56 +0200
Subject: [PATCH 08/15] Don't warn for function casts involving FARPROC

---
 gcc/c/c-typeck.c | 5 +++++
 gcc/cp/typeck.c  | 5 +++++
 2 files changed, 10 insertions(+)

diff --git a/gcc/c/c-typeck.c b/gcc/c/c-typeck.c
index 51a62c800f7..5fe5049bc15 100644
--- a/gcc/c/c-typeck.c
+++ b/gcc/c/c-typeck.c
@@ -5858,6 +5858,11 @@ c_safe_function_type_cast_p (tree t1, tree t2)
       TYPE_ARG_TYPES (t2) == void_list_node)
     return true;
 
+  /* FARPROC  */
+  if (TREE_CODE (TYPE_MAIN_VARIANT (TREE_TYPE (t2))) == INTEGER_TYPE &&
+      TYPE_ARG_TYPES (t2) == NULL_TREE)
+    return true;
+
   if (!c_safe_arg_type_equiv_p (TREE_TYPE (t1), TREE_TYPE (t2)))
     return false;
 
diff --git a/gcc/cp/typeck.c b/gcc/cp/typeck.c
index 50d0f1e6a62..31f67363a34 100644
--- a/gcc/cp/typeck.c
+++ b/gcc/cp/typeck.c
@@ -1249,6 +1249,11 @@ cxx_safe_function_type_cast_p (tree t1, tree t2)
       TYPE_ARG_TYPES (t2) == void_list_node)
     return true;
 
+  /* FARPROC  */
+  if (TREE_CODE (TYPE_MAIN_VARIANT (TREE_TYPE (t2))) == INTEGER_TYPE &&
+      TYPE_ARG_TYPES (t2) == void_list_node)
+    return true;
+
   if (!cxx_safe_arg_type_equiv_p (TREE_TYPE (t1), TREE_TYPE (t2)))
     return false;
 
-- 
2.15.1.windows.2

