From 7b1aa73d363ddd6f0cdddc3d729fcf4068c5489a Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sun, 28 Mar 2021 14:37:19 +0200
Subject: [PATCH 09/40] Fix duplicate Wformat warnings (PR c/92292)

---
 gcc/c-family/c-format.cc | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/gcc/c-family/c-format.cc b/gcc/c-family/c-format.cc
index 211d20dd25b..79dd8be366e 100644
--- a/gcc/c-family/c-format.cc
+++ b/gcc/c-family/c-format.cc
@@ -1316,6 +1316,8 @@ check_function_format (const_tree fn, tree attrs, int nargs,
 			     format_types[info.format_type].name);
 		}
 	    }
+
+	  break;
 	}
     }
 }
-- 
2.54.0

