From 01e5f8e800bf6ce622854732b7db58ce1e0ff37f Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sun, 28 Mar 2021 14:37:19 +0200
Subject: [PATCH 09/36] Fix duplicate Wformat warnings (PR c/92292)

---
 gcc/c-family/c-format.cc | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/gcc/c-family/c-format.cc b/gcc/c-family/c-format.cc
index 9c4deabc109..0c80d852b60 100644
--- a/gcc/c-family/c-format.cc
+++ b/gcc/c-family/c-format.cc
@@ -1277,6 +1277,8 @@ check_function_format (const_tree fn, tree attrs, int nargs,
 			     format_types[info.format_type].name);
 		}
 	    }
+
+	  break;
 	}
     }
 }
-- 
2.39.1.windows.1

