From 39e02938409e6aa3539380611a10cd06e33f662d Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sun, 28 Mar 2021 14:37:19 +0200
Subject: [PATCH 09/39] Fix duplicate Wformat warnings (PR c/92292)

---
 gcc/c-family/c-format.cc | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/gcc/c-family/c-format.cc b/gcc/c-family/c-format.cc
index af326fc9243..9becb919585 100644
--- a/gcc/c-family/c-format.cc
+++ b/gcc/c-family/c-format.cc
@@ -1319,6 +1319,8 @@ check_function_format (const_tree fn, tree attrs, int nargs,
 			     format_types[info.format_type].name);
 		}
 	    }
+
+	  break;
 	}
     }
 }
-- 
2.53.0

