From bd65d0707da2cb35c37f4d9f821af19a0b6afe6f Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sun, 28 Mar 2021 14:40:33 +0200
Subject: [PATCH 10/36] Enable diagnostic URLs on cygwin terminal

---
 gcc/diagnostic-color.cc | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/gcc/diagnostic-color.cc b/gcc/diagnostic-color.cc
index 91ed3057be0..2908b971265 100644
--- a/gcc/diagnostic-color.cc
+++ b/gcc/diagnostic-color.cc
@@ -266,7 +266,7 @@ w_isatty (int num)
 	  !memcmp (oni->Name.Buffer, namedPipe, l1 * 2) &&
 	  (!memcmp (oni->Name.Buffer + (nl - l2), toMaster, l2 * 2) ||
 	   !memcmp (oni->Name.Buffer + (nl - l3), toMasterNat, l3 * 2)))
-	is_a_tty = 1;
+	is_a_tty = 2;
     }
 
   free (oni);
@@ -360,7 +360,7 @@ static bool
 auto_enable_urls ()
 {
 #ifdef __MINGW32__
-  return false;
+  return w_isatty (STDERR_FILENO) == 2;
 #else
   const char *term, *colorterm;
 
-- 
2.39.1.windows.1

