From 1bdcc75fb4b21e1228ecedd5ee5fdc08875d5c18 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sun, 28 Mar 2021 14:40:33 +0200
Subject: [PATCH 10/40] Enable diagnostic URLs on cygwin terminal

---
 gcc/diagnostic-color.cc | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/gcc/diagnostic-color.cc b/gcc/diagnostic-color.cc
index aaf685fab7b..7b15ac85adb 100644
--- a/gcc/diagnostic-color.cc
+++ b/gcc/diagnostic-color.cc
@@ -323,6 +323,8 @@ w_isatty (int num)
 	  GetConsoleScreenBufferInfo (fh, &csbi);
 	  console_color = csbi.wAttributes & 0xff;
 	}
+      else
+	return 2;
       return 1;
     }
 
@@ -355,7 +357,7 @@ w_isatty (int num)
 	  !memcmp (oni->Name.Buffer, namedPipe, l1 * 2) &&
 	  (!memcmp (oni->Name.Buffer + (nl - l2), toMaster, l2 * 2) ||
 	   !memcmp (oni->Name.Buffer + (nl - l3), toMasterNat, l3 * 2)))
-	is_a_tty = 1;
+	is_a_tty = 2;
     }
 
   free (oni);
@@ -465,6 +467,9 @@ parse_env_vars_for_urls ()
 static bool
 auto_enable_urls ()
 {
+#ifdef __MINGW32__
+  return w_isatty (STDERR_FILENO) == 2;
+#else
   const char *term, *colorterm;
 
   /* First check the terminal is capable of printing color escapes,
@@ -525,6 +530,7 @@ auto_enable_urls ()
     return false;
 
   return true;
+#endif
 }
 
 /* Determine if URLs should be enabled, based on RULE,
-- 
2.54.0

