From 6c8411c063f4e15dcc645fd128047dcd8002a82b Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sat, 20 Jul 2024 20:14:09 +0200
Subject: [PATCH 15/40] Set TZDB_DISABLED if _GLIBCXX_HAS_GTHREADS is not
 available

---
 libstdc++-v3/src/c++20/tzdb.cc | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/libstdc++-v3/src/c++20/tzdb.cc b/libstdc++-v3/src/c++20/tzdb.cc
index 169810349e8..f58565c5ed1 100644
--- a/libstdc++-v3/src/c++20/tzdb.cc
+++ b/libstdc++-v3/src/c++20/tzdb.cc
@@ -92,7 +92,8 @@ namespace __gnu_cxx
 #endif
 }
 
-#if ! defined _GLIBCXX_ZONEINFO_DIR && ! defined _GLIBCXX_STATIC_TZDATA
+#if (! defined _GLIBCXX_ZONEINFO_DIR && ! defined _GLIBCXX_STATIC_TZDATA) \
+  || (USE_ATOMIC_SHARED_PTR && ! defined _GLIBCXX_HAS_GTHREADS)
 # define TZDB_DISABLED
   [[noreturn]] void __throw_disabled()
   {
-- 
2.54.0

