From 30b08a142c87f911e4cc69fb5d27fc1a3046f4c5 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sat, 20 Jul 2024 20:14:09 +0200
Subject: [PATCH 15/40] Set TZDB_DISABLED if _GLIBCXX_HAS_GTHREADS is not
 available

---
 libstdc++-v3/src/c++20/tzdb.cc | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/libstdc++-v3/src/c++20/tzdb.cc b/libstdc++-v3/src/c++20/tzdb.cc
index 6e244dc656d..b021ee85ac4 100644
--- a/libstdc++-v3/src/c++20/tzdb.cc
+++ b/libstdc++-v3/src/c++20/tzdb.cc
@@ -88,7 +88,8 @@ namespace __gnu_cxx
 #endif
 }
 
-#if ! defined _GLIBCXX_ZONEINFO_DIR && ! defined _GLIBCXX_STATIC_TZDATA
+#if (! defined _GLIBCXX_ZONEINFO_DIR && ! defined _GLIBCXX_STATIC_TZDATA) \
+  || (USE_ATOMIC_SHARED_PTR && ! defined _GLIBCXX_HAS_GTHREADS)
 # define TZDB_DISABLED
   [[noreturn]] void __throw_disabled()
   {
-- 
2.46.2.windows.1

