From 83020b256e7ba73ea5400f4cb49287a120f2f890 Mon Sep 17 00:00:00 2001
From: Ed Catmur <ed@catmur.uk>
Date: Sun, 6 Mar 2022 17:04:24 +0100
Subject: [PATCH 15/36] Add name of parameter pack

---
 gcc/dwarf2out.cc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gcc/dwarf2out.cc b/gcc/dwarf2out.cc
index 1b0e8b5a5b2..ad9f80dfc97 100644
--- a/gcc/dwarf2out.cc
+++ b/gcc/dwarf2out.cc
@@ -23140,7 +23140,7 @@ gen_formal_parameter_pack_die  (tree parm_pack,
 	      && subr_die);
 
   parm_pack_die = new_die (DW_TAG_GNU_formal_parameter_pack, subr_die, parm_pack);
-  add_src_coords_attributes (parm_pack_die, parm_pack);
+  add_name_and_src_coords_attributes (parm_pack_die, parm_pack);
 
   for (arg = pack_arg; arg; arg = DECL_CHAIN (arg))
     {
-- 
2.39.1.windows.1

