From fc7c86c8d907979843dad6a7b705e7800c8cef21 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sat, 20 Jul 2024 20:14:09 +0200
Subject: [PATCH 16/36] Set TZDB_DISABLED if _GLIBCXX_HAS_GTHREADS is not
 available

---
 libstdc++-v3/src/c++20/tzdb.cc | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/libstdc++-v3/src/c++20/tzdb.cc b/libstdc++-v3/src/c++20/tzdb.cc
index c7c7cc9deee..392187ef45e 100644
--- a/libstdc++-v3/src/c++20/tzdb.cc
+++ b/libstdc++-v3/src/c++20/tzdb.cc
@@ -85,7 +85,8 @@ namespace __gnu_cxx
 #endif
 }
 
-#if ! defined _GLIBCXX_ZONEINFO_DIR && ! defined _GLIBCXX_STATIC_TZDATA
+#if (! defined _GLIBCXX_ZONEINFO_DIR && ! defined _GLIBCXX_STATIC_TZDATA) \
+  || (USE_ATOMIC_SHARED_PTR && ! defined _GLIBCXX_HAS_GTHREADS)
 # define TZDB_DISABLED
   [[noreturn]] void __throw_disabled()
   {
-- 
2.39.1.windows.1

