From 5897343fa176d86b716d07fdef3bf12003a352ea Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Fri, 6 Sep 2024 17:38:44 +0200
Subject: [PATCH 18/39] Fix -freport-bug for Windows

---
 libiberty/pex-win32.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/libiberty/pex-win32.c b/libiberty/pex-win32.c
index 2ba9b69b830..22ac7a59fff 100644
--- a/libiberty/pex-win32.c
+++ b/libiberty/pex-win32.c
@@ -130,10 +130,9 @@ pex_win32_open_write (struct pex_obj *obj ATTRIBUTE_UNUSED, const char *name,
 {
   /* Note that we can't use O_EXCL here because gcc may have already
      created the temporary file via make_temp_file.  */
-  if (append)
-    return -1;
   return _open (name,
-		(_O_WRONLY | _O_CREAT | _O_TRUNC
+		(_O_WRONLY | _O_CREAT
+		 | (append ? _O_APPEND : _O_TRUNC)
 		 | (binary ? _O_BINARY : _O_TEXT)),
 		_S_IREAD | _S_IWRITE);
 }
-- 
2.53.0

