From d1e0790703c7557866462389beca33965f0a3248 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Fri, 5 Sep 2025 19:51:59 +0200
Subject: [PATCH 35/40] Use RVA as module offset

---
 libsanitizer/sanitizer_common/sanitizer_win.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/libsanitizer/sanitizer_common/sanitizer_win.cpp b/libsanitizer/sanitizer_common/sanitizer_win.cpp
index f3621c82044..11c65eab5b4 100644
--- a/libsanitizer/sanitizer_common/sanitizer_win.cpp
+++ b/libsanitizer/sanitizer_common/sanitizer_win.cpp
@@ -711,6 +711,7 @@ void ListOfModules::init() {
     uptr base_address = (uptr)mi.lpBaseOfDll;
     uptr end_address = (uptr)mi.lpBaseOfDll + mi.SizeOfImage;
 
+#ifndef __GNUC__
     // Adjust the base address of the module so that we get a VA instead of an
     // RVA when computing the module offset. This helps llvm-symbolizer find the
     // right DWARF CU. In the common case that the image is loaded at it's
@@ -718,6 +719,9 @@ void ListOfModules::init() {
     uptr preferred_base =
         GetPreferredBase(&module_name[0], &buf[0], buf.size());
     uptr adjusted_base = base_address - preferred_base;
+#else
+    uptr adjusted_base = base_address;
+#endif
 
     modules_.push_back(LoadedModule());
     LoadedModule &cur_module = modules_.back();
-- 
2.54.0

