From 9cd14ac0d331a0a7b2cf64bb5ee8b2ccff332d18 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Fri, 5 Sep 2025 19:51:59 +0200
Subject: [PATCH 35/39] Use RVA as module offset

---
 libsanitizer/sanitizer_common/sanitizer_win.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/libsanitizer/sanitizer_common/sanitizer_win.cpp b/libsanitizer/sanitizer_common/sanitizer_win.cpp
index fa412a58c74..3cc382ce2b7 100644
--- a/libsanitizer/sanitizer_common/sanitizer_win.cpp
+++ b/libsanitizer/sanitizer_common/sanitizer_win.cpp
@@ -726,6 +726,7 @@ void ListOfModules::init() {
     uptr base_address = (uptr)mi.lpBaseOfDll;
     uptr end_address = (uptr)mi.lpBaseOfDll + mi.SizeOfImage;
 
+#ifndef __GNUC__
     // Adjust the base address of the module so that we get a VA instead of an
     // RVA when computing the module offset. This helps llvm-symbolizer find the
     // right DWARF CU. In the common case that the image is loaded at it's
@@ -733,6 +734,9 @@ void ListOfModules::init() {
     uptr preferred_base =
         GetPreferredBase(&module_name[0], &buf[0], buf.size());
     uptr adjusted_base = base_address - preferred_base;
+#else
+    uptr adjusted_base = base_address;
+#endif
 
     modules_.push_back(LoadedModule());
     LoadedModule &cur_module = modules_.back();
-- 
2.53.0

