From 634074b10923234823efd857312259aef2a9fd52 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Mon, 8 Sep 2025 15:55:08 +0200
Subject: [PATCH 36/39] Align the stack on new threads

---
 libsanitizer/asan/asan_win.cpp | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/libsanitizer/asan/asan_win.cpp b/libsanitizer/asan/asan_win.cpp
index aae49fd5388..abd998f9eda 100644
--- a/libsanitizer/asan/asan_win.cpp
+++ b/libsanitizer/asan/asan_win.cpp
@@ -138,6 +138,9 @@ struct ThreadStartParams {
   void *arg;
 };
 
+#ifndef _WIN64
+__attribute__((force_align_arg_pointer))
+#endif
 static thread_return_t THREAD_CALLING_CONV asan_thread_start(void *arg) {
   AsanThread *t = (AsanThread *)arg;
   SetCurrentThread(t);
-- 
2.53.0

