--- gcc/common.opt	2016-06-26 15:07:45.919372100 +0200
+++ gcc/common.opt	2016-06-28 19:22:10.918590100 +0200
@@ -2686,6 +2686,10 @@
 Common JoinedOrMissing Negative(gdwarf)
 Generate debug information in COFF format.
 
+gcolumn-info
+Common Report Var(flag_gcolumn_info)
+Emit column location information.
+
 gdwarf
 Common JoinedOrMissing Negative(gdwarf-)
 Generate debug information in default version of DWARF format.
--- gcc/debug.h	2016-01-04 15:30:50.000000000 +0100
+++ gcc/debug.h	2016-06-28 19:43:36.486120400 +0200
@@ -67,11 +67,13 @@
 
   /* Record a source file location at (FILE, LINE, DISCRIMINATOR).  */
   void (* source_line) (unsigned int line, const char *file,
-                        int discriminator, bool is_stmt);
+                        int discriminator, bool is_stmt,
+                        unsigned int column);
 
   /* Called at start of prologue code.  LINE is the first line in the
      function.  */
-  void (* begin_prologue) (unsigned int line, const char *file);
+  void (* begin_prologue) (unsigned int line, const char *file,
+                           unsigned int column);
 
   /* Called at end of prologue code.  LINE is the first line in the
      function.  */
@@ -194,8 +196,10 @@
 extern void debug_nothing_void (void);
 extern void debug_nothing_charstar (const char *);
 extern void debug_nothing_int_charstar (unsigned int, const char *);
+extern void debug_nothing_int_charstar_int (unsigned int, const char *,
+                                            unsigned int);
 extern void debug_nothing_int_charstar_int_bool (unsigned int, const char *,
-                                                 int, bool);
+                                                 int, bool, unsigned int);
 extern void debug_nothing_int (unsigned int);
 extern void debug_nothing_int_int (unsigned int, unsigned int);
 extern void debug_nothing_tree (tree);
@@ -217,7 +221,7 @@
 
 /* Dwarf2 frame information.  */
 
-extern void dwarf2out_begin_prologue (unsigned int, const char *);
+extern void dwarf2out_begin_prologue (unsigned int, const char *, unsigned int);
 extern void dwarf2out_vms_end_prologue (unsigned int, const char *);
 extern void dwarf2out_vms_begin_epilogue (unsigned int, const char *);
 extern void dwarf2out_end_epilogue (unsigned int, const char *);
--- gcc/debug.c	2016-01-04 15:30:50.000000000 +0100
+++ gcc/debug.c	2016-06-28 19:43:36.492120700 +0200
@@ -36,7 +36,7 @@
   debug_nothing_int_int,	         /* end_block */
   debug_true_const_tree,	         /* ignore_block */
   debug_nothing_int_charstar_int_bool,	 /* source_line */
-  debug_nothing_int_charstar,	         /* begin_prologue */
+  debug_nothing_int_charstar_int,	 /* begin_prologue */
   debug_nothing_int_charstar,	         /* end_prologue */
   debug_nothing_int_charstar,	         /* begin_epilogue */
   debug_nothing_int_charstar,	         /* end_epilogue */
@@ -115,10 +115,18 @@
 }
 
 void
+debug_nothing_int_charstar_int (unsigned int line ATTRIBUTE_UNUSED,
+				const char *text ATTRIBUTE_UNUSED,
+				unsigned int column ATTRIBUTE_UNUSED)
+{
+}
+
+void
 debug_nothing_int_charstar_int_bool (unsigned int line ATTRIBUTE_UNUSED,
 			             const char *text ATTRIBUTE_UNUSED,
 			             int discriminator ATTRIBUTE_UNUSED,
-			             bool is_stmt ATTRIBUTE_UNUSED)
+			             bool is_stmt ATTRIBUTE_UNUSED,
+				     unsigned int column ATTRIBUTE_UNUSED)
 {
 }
 
--- gcc/dwarf2out.c	2016-03-17 14:43:01.000000000 +0100
+++ gcc/dwarf2out.c	2016-06-28 20:09:51.138185300 +0200
@@ -92,7 +92,7 @@
 #include "gdb/gdb-index.h"
 #include "rtl-iter.h"
 
-static void dwarf2out_source_line (unsigned int, const char *, int, bool);
+static void dwarf2out_source_line (unsigned int, const char *, int, bool, unsigned int);
 static rtx_insn *last_var_location_insn;
 static rtx_insn *cached_next_real_insn;
 static void dwarf2out_decl (tree);
@@ -1031,7 +1031,8 @@
 
 void
 dwarf2out_begin_prologue (unsigned int line ATTRIBUTE_UNUSED,
-			  const char *file ATTRIBUTE_UNUSED)
+			  const char *file ATTRIBUTE_UNUSED,
+			  unsigned int column ATTRIBUTE_UNUSED)
 {
   char label[MAX_ARTIFICIAL_LABEL_BYTES];
   char * dup_label;
@@ -1081,7 +1082,7 @@
      prologue case, not the eh frame case.  */
 #ifdef DWARF2_DEBUGGING_INFO
   if (file)
-    dwarf2out_source_line (line, file, 0, true);
+    dwarf2out_source_line (line, file, 0, true, column);
 #endif
 
   if (dwarf2out_do_cfi_asm ())
@@ -2569,7 +2570,7 @@
   debug_nothing_int_int,	         /* end_block */
   debug_true_const_tree,	         /* ignore_block */
   dwarf2out_source_line,	 /* source_line */
-  debug_nothing_int_charstar,	         /* begin_prologue */
+  debug_nothing_int_charstar_int,	 /* begin_prologue */
   debug_nothing_int_charstar,	         /* end_prologue */
   debug_nothing_int_charstar,	         /* begin_epilogue */
   debug_nothing_int_charstar,	         /* end_epilogue */
@@ -21326,6 +21327,8 @@
     {
       add_AT_file (die, DW_AT_call_file, lookup_filename (s.file));
       add_AT_unsigned (die, DW_AT_call_line, s.line);
+      if (flag_gcolumn_info)
+	add_AT_unsigned (die, DW_AT_call_column, s.column);
     }
 }
 
@@ -24527,6 +24530,7 @@
   table = ggc_cleared_alloc<dw_line_info_table> ();
   table->file_num = 1;
   table->line_num = 1;
+  table->column_num = 0;
   table->is_stmt = DWARF_LINE_DEFAULT_IS_STMT_START;
 
   return table;
@@ -24703,7 +24707,7 @@
 
 static void
 dwarf2out_source_line (unsigned int line, const char *filename,
-                       int discriminator, bool is_stmt)
+                       int discriminator, bool is_stmt, unsigned int column)
 {
   unsigned int file_num;
   dw_line_info_table *table;
@@ -24733,8 +24737,12 @@
      insn that corresponds to something the user wrote.  These may be
      very different locations once scheduling is enabled.  */
 
+  if (!flag_gcolumn_info)
+    column = 0;
+
   if (0 && file_num == table->file_num
       && line == table->line_num
+      && column == table->column_num
       && discriminator == table->discrim_num
       && is_stmt == table->is_stmt)
     return;
@@ -24755,7 +24763,7 @@
       putc (' ', asm_out_file);
       fprint_ul (asm_out_file, line);
       putc (' ', asm_out_file);
-      putc ('0', asm_out_file);
+      fprint_ul (asm_out_file, column);
 
       if (is_stmt != table->is_stmt)
 	{
@@ -24783,11 +24791,14 @@
 	push_dw_line_info_entry (table, LI_set_discriminator, discriminator);
       if (is_stmt != table->is_stmt)
 	push_dw_line_info_entry (table, LI_negate_stmt, 0);
+      if (column)
+	push_dw_line_info_entry (table, LI_set_column, column);
       push_dw_line_info_entry (table, LI_set_line, line);
     }
 
   table->file_num = file_num;
   table->line_num = line;
+  table->column_num = column;
   table->discrim_num = discriminator;
   table->is_stmt = is_stmt;
   table->in_use = true;
--- gcc/dbxout.c	2016-01-04 15:30:50.000000000 +0100
+++ gcc/dbxout.c	2016-06-28 19:43:36.521122400 +0200
@@ -331,8 +331,9 @@
 /* The debug hooks structure.  */
 #if defined (DBX_DEBUGGING_INFO)
 
-static void dbxout_source_line (unsigned int, const char *, int, bool);
-static void dbxout_begin_prologue (unsigned int, const char *);
+static void dbxout_source_line (unsigned int, const char *, int, bool,
+                                unsigned int);
+static void dbxout_begin_prologue (unsigned int, const char *, unsigned int);
 static void dbxout_source_file (const char *);
 static void dbxout_function_end (tree);
 static void dbxout_begin_function (tree);
@@ -1240,7 +1241,8 @@
    function scope  */
 
 static void
-dbxout_begin_prologue (unsigned int lineno, const char *filename)
+dbxout_begin_prologue (unsigned int lineno, const char *filename,
+                       unsigned int column ATTRIBUTE_UNUSED)
 {
   if (use_gnu_debug_info_extensions
       && !NO_DBX_FUNCTION_END
@@ -1251,7 +1253,7 @@
   /* pre-increment the scope counter */
   scope_labelno++;
 
-  dbxout_source_line (lineno, filename, 0, true);
+  dbxout_source_line (lineno, filename, 0, true, 0);
   /* Output function begin block at function scope, referenced
      by dbxout_block, dbxout_source_line and dbxout_function_end.  */
   emit_pending_bincls_if_required ();
@@ -1264,7 +1266,8 @@
 static void
 dbxout_source_line (unsigned int lineno, const char *filename,
                     int discriminator ATTRIBUTE_UNUSED,
-                    bool is_stmt ATTRIBUTE_UNUSED)
+                    bool is_stmt ATTRIBUTE_UNUSED,
+                    unsigned int column ATTRIBUTE_UNUSED)
 {
   dbxout_source_file (filename);
 
--- gcc/sdbout.c	2016-01-04 15:30:50.000000000 +0100
+++ gcc/sdbout.c	2016-06-28 19:48:14.953047800 +0200
@@ -115,11 +115,13 @@
 static void sdbout_end_source_file	(unsigned int);
 static void sdbout_begin_block		(unsigned int, unsigned int);
 static void sdbout_end_block		(unsigned int, unsigned int);
-static void sdbout_source_line		(unsigned int, const char *, int, bool);
+static void sdbout_source_line		(unsigned int, const char *, int, bool,
+					 unsigned int);
 static void sdbout_end_epilogue		(unsigned int, const char *);
 static void sdbout_early_global_decl	(tree);
 static void sdbout_late_global_decl	(tree);
-static void sdbout_begin_prologue	(unsigned int, const char *);
+static void sdbout_begin_prologue	(unsigned int, const char *,
+					 unsigned int);
 static void sdbout_end_prologue		(unsigned int, const char *);
 static void sdbout_begin_function	(tree);
 static void sdbout_end_function		(unsigned int);
@@ -1522,7 +1524,8 @@
 static void
 sdbout_source_line (unsigned int line, const char *filename ATTRIBUTE_UNUSED,
                     int discriminator ATTRIBUTE_UNUSED,
-                    bool is_stmt ATTRIBUTE_UNUSED)
+                    bool is_stmt ATTRIBUTE_UNUSED,
+                    unsigned int column ATTRIBUTE_UNUSED)
 {
   /* COFF relative line numbers must be positive.  */
   if ((int) line > sdb_begin_function_line)
@@ -1552,7 +1555,8 @@
    describe the parameter list.  */
 
 static void
-sdbout_begin_prologue (unsigned int line, const char *file ATTRIBUTE_UNUSED)
+sdbout_begin_prologue (unsigned int line, const char *file ATTRIBUTE_UNUSED,
+                       unsigned int column ATTRIBUTE_UNUSED)
 {
   sdbout_end_prologue (line, file);
 }
--- gcc/final.c	2016-01-22 17:44:10.000000000 +0100
+++ gcc/final.c	2016-06-28 20:12:31.475356100 +0200
@@ -117,6 +117,9 @@
 /* Line number of last NOTE.  */
 static int last_linenum;
 
+/* Column number of last NOTE.  */
+static int last_column;
+
 /* Last discriminator written to assembly.  */
 static int last_discriminator;
 
@@ -135,6 +138,7 @@
 /* Override filename and line number.  */
 static const char *override_filename;
 static int override_linenum;
+static int override_column;
 
 /* Whether to force emission of a line note before the next insn.  */
 static bool force_source_line = false;
@@ -1762,6 +1766,7 @@
 
   last_filename = LOCATION_FILE (prologue_location);
   last_linenum = LOCATION_LINE (prologue_location);
+  last_column = flag_gcolumn_info ? LOCATION_COLUMN (prologue_location) : 0;
   last_discriminator = discriminator = 0;
 
   high_block_linenum = high_function_linenum = last_linenum;
@@ -1770,10 +1775,10 @@
     asan_function_start ();
 
   if (!DECL_IGNORED_P (current_function_decl))
-    debug_hooks->begin_prologue (last_linenum, last_filename);
+    debug_hooks->begin_prologue (last_linenum, last_filename, last_column);
 
   if (!dwarf2_debug_info_emitted_p (current_function_decl))
-    dwarf2out_begin_prologue (0, NULL);
+    dwarf2out_begin_prologue (0, NULL, 0);
 
 #ifdef LEAF_REG_REMAP
   if (crtl->uses_only_leaf_regs)
@@ -2311,6 +2316,7 @@
 		{
 		  override_filename = LOCATION_FILE (*locus_ptr);
 		  override_linenum = LOCATION_LINE (*locus_ptr);
+		  override_column = LOCATION_COLUMN (*locus_ptr);
 		}
 	    }
 	  break;
@@ -2344,11 +2350,13 @@
 		{
 		  override_filename = LOCATION_FILE (*locus_ptr);
 		  override_linenum = LOCATION_LINE (*locus_ptr);
+		  override_column = LOCATION_COLUMN (*locus_ptr);
 		}
 	      else
 		{
 		  override_filename = NULL;
 		  override_linenum = 0;
+		  override_column = 0;
 		}
 	    }
 	  break;
@@ -2564,7 +2572,8 @@
 	if (!DECL_IGNORED_P (current_function_decl)
 	    && notice_source_line (insn, &is_stmt))
 	  (*debug_hooks->source_line) (last_linenum, last_filename,
-				       last_discriminator, is_stmt);
+				       last_discriminator, is_stmt,
+				       last_column);
 
 	if (GET_CODE (body) == ASM_INPUT)
 	  {
@@ -3045,22 +3054,26 @@
 {
   const char *filename;
   int linenum;
+  int column;
 
   if (override_filename)
     {
       filename = override_filename;
       linenum = override_linenum;
+      column = flag_gcolumn_info ? override_column : 0;
     }
   else if (INSN_HAS_LOCATION (insn))
     {
       expanded_location xloc = insn_location (insn);
       filename = xloc.file;
       linenum = xloc.line;
+      column = flag_gcolumn_info ? xloc.column : 0;
     }
   else
     {
       filename = NULL;
       linenum = 0;
+      column = 0;
     }
 
   if (filename == NULL)
@@ -3068,11 +3081,13 @@
 
   if (force_source_line
       || filename != last_filename
-      || last_linenum != linenum)
+      || last_linenum != linenum
+      || last_column != column)
     {
       force_source_line = false;
       last_filename = filename;
       last_linenum = linenum;
+      last_column = column;
       last_discriminator = discriminator;
       *is_stmt = true;
       high_block_linenum = MAX (last_linenum, high_block_linenum);
