From 70c6c27143b4f15f98fb88cce1708ef5ebd11c0e Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Tue, 4 Mar 2014 19:43:44 +0100
Subject: [PATCH 02/14] Add install/uninstall commands for JIT debugger

---
 gdb/windows-nat.c | 65 +++++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 65 insertions(+)

diff --git a/gdb/windows-nat.c b/gdb/windows-nat.c
index a0db993..700eab4 100644
--- a/gdb/windows-nat.c
+++ b/gdb/windows-nat.c
@@ -38,6 +38,7 @@
 #include <windows.h>
 #include <imagehlp.h>
 #include <psapi.h>
+#include <inttypes.h>
 #ifdef __CYGWIN__
 #include <wchar.h>
 #include <sys/cygwin.h>
@@ -982,6 +983,66 @@ signal_event_command (char *args, int from_tty)
   CloseHandle ((HANDLE) event_id);
 }
 
+static void
+jit_install_command (char *args, int from_tty)
+{
+  char gdb_file[MAX_PATH];
+  char command_str[MAX_PATH + 100];
+  HKEY hKey;
+  LONG ret;
+
+  dont_repeat ();
+
+  if (!GetModuleFileNameA(NULL, gdb_file, MAX_PATH))
+    error (_("Error getting gdb name: %u."),
+	   (unsigned) GetLastError ());
+
+  strcpy (command_str, "\"");
+  strcat (command_str, gdb_file);
+  strcat (command_str, "\" -q -p %ld -ex \"signal-event %ld\" -ex \"c\"");
+
+  if (RegCreateKeyEx (HKEY_LOCAL_MACHINE,
+		      "Software\\Microsoft\\Windows NT\\"
+		      "CurrentVersion\\AeDebug",
+		      0, NULL, REG_OPTION_NON_VOLATILE, KEY_WRITE, NULL,
+		      &hKey, NULL) != ERROR_SUCCESS)
+    error (_("Error creating 'AeDebug' registry key."));
+
+  ret = RegSetValueEx (hKey, "Debugger", 0, REG_SZ,
+			(const BYTE *)command_str,
+			strlen (command_str));
+  if (ret == ERROR_SUCCESS)
+    RegSetValueEx (hKey, "Auto", 0, REG_SZ, (const BYTE *)"1", 1);
+
+  RegCloseKey (hKey);
+
+  if (ret != ERROR_SUCCESS)
+    error (_("Error setting 'Debugger' registry value."));
+}
+
+static void
+jit_uninstall_command (char *args, int from_tty)
+{
+  HKEY hKey;
+  LONG ret;
+
+  dont_repeat ();
+
+  if (RegOpenKeyEx (HKEY_LOCAL_MACHINE,
+		    "Software\\Microsoft\\Windows NT\\"
+		    "CurrentVersion\\AeDebug",
+		    0, KEY_ALL_ACCESS, &hKey) != ERROR_SUCCESS)
+    error (_("Error opening 'AeDebug' registry key."));
+
+  ret = RegDeleteValue (hKey, "Debugger");
+  RegSetValueEx (hKey, "Auto", 0, REG_SZ, (const BYTE *)"0", 1);
+
+  RegCloseKey (hKey);
+
+  if (ret != ERROR_SUCCESS)
+    error (_("Error deleting 'Debugger' registry value."));
+}
+
 /* Handle DEBUG_STRING output from child process.
    Cygwin prepends its messages with a "cygwin:".  Interpret this as
    a Cygwin signal.  Otherwise just print the string as a warning.  */
@@ -2701,6 +2762,10 @@ _initialize_windows_nat (void)
 
   c = add_com ("signal-event", class_run, signal_event_command,
 	       _("Signal an object with ID."));
+  c = add_com ("jit-install", class_obscure, jit_install_command,
+	       _("Install as JIT debugger."));
+  c = add_com ("jit-uninstall", class_obscure, jit_uninstall_command,
+	       _("Uninstall JIT debugger."));
 
   c = add_com ("dll-symbols", class_files, dll_symbol_command,
 	       _("Load dll library symbols from FILE."));
-- 
1.8.5.2.msysgit.0

