From b4bf3598b6c0736aa13a1a7e9509df458fa634f0 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Tue, 4 Mar 2014 19:53:04 +0100
Subject: [PATCH 03/15] Only build missing texi files

---
 gdb/doc/Makefile.in | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/gdb/doc/Makefile.in b/gdb/doc/Makefile.in
index c9c8c85..91f021b 100644
--- a/gdb/doc/Makefile.in
+++ b/gdb/doc/Makefile.in
@@ -135,7 +135,7 @@ GDB_DOC_FILES = \
 	$(srcdir)/gdb.texinfo \
 	$(srcdir)/guile.texi \
 	$(srcdir)/python.texi \
-	$(GDB_DOC_SOURCE_INCLUDES) \
+	$(GDB_DOC_SOURCE_INCLUDES) | \
 	$(GDB_DOC_BUILD_INCLUDES)
 
 # Stabs manual: All files
@@ -144,7 +144,7 @@ STABS_DOC_SOURCE_INCLUDES = \
 STABS_DOC_BUILD_INCLUDES =
 STABS_DOC_FILES = \
 	$(srcdir)/stabs.texinfo \
-	$(STABS_DOC_SOURCE_INCLUDES) \
+	$(STABS_DOC_SOURCE_INCLUDES) | \
 	$(STABS_DOC_BUILD_INCLUDES)
 
 # Annotate migration document
@@ -155,7 +155,7 @@ ANNOTATE_DOC_BUILD_INCLUDES = \
 	GDBvn.texi
 ANNOTATE_DOC_FILES = \
 	$(srcdir)/annotate.texinfo \
-	$(ANNOTATE_DOC_SOURCE_INCLUDES) \
+	$(ANNOTATE_DOC_SOURCE_INCLUDES) | \
 	$(ANNOTATE_DOC_BUILD_INCLUDES)
 
 # Options to extract the man page from gdb.texinfo
-- 
1.9.4.msysgit.2

