--- a/bfd/doc/Makefile.in	2015-01-15 11:58:11.000000000 +0100
+++ b/bfd/doc/Makefile.in	2015-07-29 20:42:40.472068900 +0200
@@ -436,7 +436,7 @@
 clean-libtool:
 	-rm -rf .libs _libs
 
-bfd.info: bfd.texinfo $(bfd_TEXINFOS)
+bfd.info: bfd.texinfo | $(bfd_TEXINFOS)
 	restore=: && backupdir="$(am__leading_dot)am$$$$" && \
 	rm -rf $$backupdir && mkdir $$backupdir && \
 	if ($(MAKEINFO) --version) >/dev/null 2>&1; then \
--- a/gdb/doc/Makefile.in	2015-01-15 11:58:12.000000000 +0100
+++ b/gdb/doc/Makefile.in	2015-07-29 20:50:55.494382600 +0200
@@ -144,7 +144,7 @@
 STABS_DOC_BUILD_INCLUDES =
 STABS_DOC_FILES = \
 	$(srcdir)/stabs.texinfo \
-	$(STABS_DOC_SOURCE_INCLUDES) \
+	$(STABS_DOC_SOURCE_INCLUDES) | \
 	$(STABS_DOC_BUILD_INCLUDES)
 
 # Annotate migration document
@@ -155,7 +155,7 @@
 	GDBvn.texi
 ANNOTATE_DOC_FILES = \
 	$(srcdir)/annotate.texinfo \
-	$(ANNOTATE_DOC_SOURCE_INCLUDES) \
+	$(ANNOTATE_DOC_SOURCE_INCLUDES) | \
 	$(ANNOTATE_DOC_BUILD_INCLUDES)
 
 # Options to extract the man page from gdb.texinfo
@@ -481,7 +481,7 @@
 		$(srcdir)/gdb.texinfo
 
 # GDB MANUAL: info file
-gdb.info: ${GDB_DOC_FILES}
+gdb.info: | ${GDB_DOC_FILES}
 	$(MAKEINFO_CMD) $(READLINE_TEXI_INCFLAG) -I ${GDBMI_DIR} -I $(srcdir) \
 		-o gdb.info $(srcdir)/gdb.texinfo
 
