From d71cd6c7a0b2752bc3c4bf118ac34d5863d4b0ac Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Tue, 4 Mar 2014 19:54:46 +0100
Subject: [PATCH 04/15] Fix mingw build with curses

---
 gdb/configure    | 4 +++-
 gdb/configure.ac | 4 +++-
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/gdb/configure b/gdb/configure
index 4b2906e..c4e5dfb 100755
--- a/gdb/configure
+++ b/gdb/configure
@@ -7111,7 +7111,9 @@ case $host_os in
     ;;
   *mingw32*)
     ac_cv_search_tgetent="none required"
-    CONFIG_OBS="$CONFIG_OBS windows-termcap.o"
+    if test "$curses_found" = no; then
+      CONFIG_OBS="$CONFIG_OBS windows-termcap.o"
+    fi
     ;;
 esac
 
diff --git a/gdb/configure.ac b/gdb/configure.ac
index c083e52..8e5525d 100644
--- a/gdb/configure.ac
+++ b/gdb/configure.ac
@@ -628,7 +628,9 @@ case $host_os in
     ;;
   *mingw32*)
     ac_cv_search_tgetent="none required"
-    CONFIG_OBS="$CONFIG_OBS windows-termcap.o"
+    if test "$curses_found" = no; then
+      CONFIG_OBS="$CONFIG_OBS windows-termcap.o"
+    fi
     ;;
 esac
 
-- 
1.9.4.msysgit.2

