From aa5ba392b51331f1a0c9587b2b61148d63e2a23f Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Wed, 12 Mar 2014 18:57:12 +0100
Subject: [PATCH 11/23] Fix search for TUI

---
 gdb/source.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/gdb/source.c b/gdb/source.c
index f48546f..248ff12 100644
--- a/gdb/source.c
+++ b/gdb/source.c
@@ -45,6 +45,10 @@
 #include "common/enum-flags.h"
 #include <algorithm>
 
+#ifdef TUI
+#include "tui/tui.h"
+#endif
+
 #define OPEN_MODE (O_RDONLY | O_BINARY)
 #define FDOPEN_MODE FOPEN_RB
 
@@ -161,6 +165,10 @@ get_first_line_listed (void)
 static void
 clear_lines_listed_range (void)
 {
+#ifdef TUI
+  if (tui_active) return;
+#endif
+
   first_line_listed = 0;
   last_line_listed = 0;
 }
@@ -1469,6 +1477,7 @@ print_source_lines_base (struct symtab *s, int line, int stopline,
   current_source_symtab = s;
   current_source_line = line;
   first_line_listed = line;
+  last_line_listed = line;
 
   /* If printing of source lines is disabled, just print file and line
      number.  */
@@ -1858,7 +1867,7 @@ reverse_search_command (char *regex, int from_tty)
   stream = fdopen (desc, FDOPEN_MODE);
   clearerr (stream);
   cleanups = make_cleanup_fclose (stream);
-  while (line > 1)
+  while (1)
     {
 /* FIXME!!!  We walk right off the end of buf if we get a long line!!!  */
       char buf[4096];		/* Should be reasonable???  */
@@ -1893,6 +1902,7 @@ reverse_search_command (char *regex, int from_tty)
 	  return;
 	}
       line--;
+      if (line < 1) break;
       if (fseek (stream, current_source_symtab->line_charpos[line - 1], 0) < 0)
 	{
 	  const char *filename;
-- 
1.9.4.msysgit.2

