From e7a6970939a0988f7f9db29d8b4d1713a239dd77 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Wed, 5 Nov 2014 19:58:33 +0100
Subject: [PATCH 14/15] Use %USERPROFILE% as alternative to %HOME%

---
 gdb/main.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/gdb/main.c b/gdb/main.c
index 3a4e402..7b54349 100644
--- a/gdb/main.c
+++ b/gdb/main.c
@@ -254,6 +254,10 @@ get_init_files (const char **system_gdbinit,
 	}
 
       homedir = getenv ("HOME");
+#ifdef __MINGW32__
+      if (!homedir)
+	homedir = getenv ("USERPROFILE");
+#endif
 
       /* If the .gdbinit file in the current directory is the same as
 	 the $HOME/.gdbinit file, it should not be sourced.  homebuf
--- a/gdb/windows-nat.c	2014-11-05 20:12:00.554412000 +0100
+++ b/gdb/windows-nat.c	2015-01-01 17:19:16.748732900 +0100
@@ -2999,6 +2999,10 @@
     return;
 
   homedir = getenv ("HOME");
+#ifdef __MINGW32__
+  if (!homedir)
+    homedir = getenv ("USERPROFILE");
+#endif
   if (homedir)
     {
       char *p;
--- a/readline/tilde.c	2015-01-01 17:20:55.628871300 +0100
+++ b/readline/tilde.c	2015-01-01 17:21:01.898880100 +0100
@@ -366,6 +366,11 @@
       if (expansion == 0)
 	expansion = sh_get_home_dir ();
 
+#ifdef __MINGW32__
+      if (expansion == 0)
+	expansion = getenv ("USERPROFILE");
+#endif
+
       return (glue_prefix_and_suffix (expansion, filename, 1));
     }
 
-- 
1.9.4.msysgit.2

