From 1348ff2d1cf6ac51288d01e6048a022bc7837f3e Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Wed, 14 Jan 2015 19:17:42 +0100
Subject: [PATCH 15/23] Don't warn for debuglink section

---
 gdb/symfile.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/gdb/symfile.c b/gdb/symfile.c
index 846aabe..1093a46 100644
--- a/gdb/symfile.c
+++ b/gdb/symfile.c
@@ -679,6 +679,7 @@ addr_info_make_relative (struct section_addr_info *addrs, bfd *abfd)
 
 	  if (!(strcmp (sect_name, ".gnu.liblist") == 0
 		|| strcmp (sect_name, ".gnu.conflict") == 0
+		|| strcmp (sect_name, ".gnu_debuglink") == 0
 		|| (strcmp (sect_name, ".bss") == 0
 		    && i > 0
 		    && strcmp (addrs->other[i - 1].name, ".dynbss") == 0
-- 
1.9.4.msysgit.2

