--- a/gdb/tui/tui-data.h	2015-01-09 17:05:18.603901600 +0100
+++ b/gdb/tui/tui-data.h	2015-01-09 17:05:25.319285700 +0100
@@ -207,6 +207,7 @@
   char full_name[MAX_LOCATOR_ELEMENT_LEN];
   char proc_name[MAX_LOCATOR_ELEMENT_LEN];
   int line_no;
+  char line_file_name[MAX_LOCATOR_ELEMENT_LEN];
   CORE_ADDR addr;
   /* Architecture associated with code at this location.  */
   struct gdbarch *gdbarch;
--- a/gdb/tui/tui-data.c	2015-01-09 17:10:16.813958200 +0100
+++ b/gdb/tui/tui-data.c	2015-01-09 17:11:14.259243900 +0100
@@ -515,6 +515,7 @@
       element->which_element.locator.full_name[0] =
 	element->which_element.locator.proc_name[0] = (char) 0;
       element->which_element.locator.line_no = 0;
+      element->which_element.locator.line_file_name[0] = (char) 0;
       element->which_element.locator.addr = 0;
       break;
     case EXEC_INFO_WIN:
--- a/gdb/tui/tui-stack.c	2015-01-09 17:11:48.462200200 +0100
+++ b/gdb/tui/tui-stack.c	2015-01-09 17:11:55.640610800 +0100
@@ -317,6 +317,8 @@
   element->proc_name[0] = (char) 0;
   strcat_to_buf (element->proc_name, MAX_LOCATOR_ELEMENT_LEN, procname);
   element->line_no = lineno;
+  element->line_file_name[0] = 0;
+  strcat_to_buf (element->line_file_name, MAX_LOCATOR_ELEMENT_LEN, fullname);
   element->addr = addr;
   element->gdbarch = gdbarch;
   tui_set_locator_fullname (fullname);
--- a/gdb/tui/tui-source.c	2015-01-09 17:12:58.977233400 +0100
+++ b/gdb/tui/tui-source.c	2015-01-09 17:13:08.188760300 +0100
@@ -679,7 +679,7 @@
 			cur_line_no;
 		      element->which_element.source.is_exec_point =
 			(filename_cmp (((struct tui_win_element *)
-				       locator->content[0])->which_element.locator.full_name,
+				       locator->content[0])->which_element.locator.line_file_name,
 				       symtab_to_fullname (s)) == 0
 			 && cur_line_no == ((struct tui_win_element *)
 					    locator->content[0])->which_element.locator.line_no);
