--- a/gdb/tui/tui-source.c	2015-01-15 17:18:54.057264000 +0100
+++ b/gdb/tui/tui-source.c	2015-02-12 17:28:12.307256300 +0100
@@ -859,8 +859,13 @@
 	  for (i = xpos + warning_length; i < line_width; i++)
 	    src_line[i] = ' ';
 
-	  src_line[i] = '\n';
+	  src_line[i] = (char) 0;
 
+#ifdef TUI_SYNTAX_HIGHLIGHT
+	  memset (src_line + line_width, 0, line_width);
+	  src_line[line_width + xpos] = COL_LITERAL;
+	  src_line[line_width + xpos + warning_length - 1] = COL_LITERAL;
+#endif
 	}			/* end if */
 
       curr_line++;
