--- a/gdb/tui/tui.c	2016-03-06 14:02:34.298887700 +0100
+++ b/gdb/tui/tui.c	2016-08-18 14:38:22.881022900 +0200
@@ -622,6 +622,8 @@
   rl_startup_hook = 0;
   rl_already_prompted = 0;
 
+  mouse_set (0);
+
   /* Leave curses and restore previous gdb terminal setting.  */
   endwin ();
 
--- a/gdb/tui/tui-io.c	2016-03-08 20:40:36.793009400 +0100
+++ b/gdb/tui/tui-io.c	2016-03-08 20:40:42.308324800 +0100
@@ -295,6 +295,8 @@
      rl_prompt because it points to an alloca buffer).  */
   xfree (tui_rl_saved_prompt);
   tui_rl_saved_prompt = rl_prompt != NULL ? xstrdup (rl_prompt) : NULL;
+
+  mouse_set (MOUSE_WHEEL_SCROLL);
 }
 
 /* Readline callback to restore the terminal.  It is called once each
@@ -302,6 +304,7 @@
 static void
 tui_deprep_terminal (void)
 {
+  mouse_set (0);
 }
 
 #ifdef TUI_USE_PIPE_FOR_READLINE
--- a/gdb/ser-mingw.c	2016-02-10 04:19:39.000000000 +0100
+++ b/gdb/ser-mingw.c	2016-03-08 19:50:48.217072600 +0100
@@ -605,6 +605,12 @@
 		  break;
 		}
 	    }
+	  else if (record.EventType == MOUSE_EVENT &&
+		   record.Event.MouseEvent.dwEventFlags == MOUSE_WHEELED)
+	    {
+	      SetEvent (state->read_event);
+	      break;
+	    }
 
 	  /* Otherwise discard it and wait again.  */
 	  ReadConsoleInput (h, &record, 1, &n_records);
--- a/gdb/tui/tui-command.c	2016-02-10 04:19:39.000000000 +0100
+++ b/gdb/tui/tui-command.c	2016-03-08 20:09:08.231989800 +0100
@@ -76,6 +76,13 @@
     case KEY_LEFT:
       tui_scroll_right (win_info, 1);
       break;
+    case KEY_MOUSE:
+      request_mouse_pos ();
+      if (MOUSE_WHEEL_UP)
+	tui_scroll_backward (win_info, 3);
+      else if (MOUSE_WHEEL_DOWN)
+	tui_scroll_forward (win_info, 3);
+      break;
     case '\f':
       break;
     default:
