--- a/gdb/windows-nat.c	2016-04-22 17:22:28.815067500 +0200
+++ b/gdb/windows-nat.c	2016-04-25 17:26:23.441670000 +0200
@@ -446,6 +446,8 @@
     {
       windows_thread_info *here = th->next;
       th->next = here->next;
+      if (here->name)
+	xfree (here->name);
       xfree (here);
     }
 }
@@ -1199,6 +1201,17 @@
     host_address_to_string (\
       current_event.u.Exception.ExceptionRecord.ExceptionAddress))
 
+#pragma pack(push,8)
+typedef struct tagTHREADNAME_INFO
+{
+  DWORD dwType;
+  LPCSTR szName;
+  DWORD dwThreadID;
+  DWORD dwFlags;
+} THREADNAME_INFO;
+#pragma pack(pop)
+#define EXCEPTION_THREAD_NAME 0x406d1388
+
 static int
 handle_exception (struct target_waitstatus *ourstatus)
 {
@@ -1318,6 +1331,37 @@
       DEBUG_EXCEPTION_SIMPLE ("EXCEPTION_NONCONTINUABLE_EXCEPTION");
       ourstatus->value.sig = GDB_SIGNAL_ILL;
       break;
+    case EXCEPTION_THREAD_NAME:
+      if (current_event.u.Exception.ExceptionRecord.NumberParameters
+	  == sizeof (THREADNAME_INFO) / sizeof (ULONG_PTR))
+	{
+	  THREADNAME_INFO *tni = (THREADNAME_INFO *)
+	    current_event.u.Exception.ExceptionRecord.ExceptionInformation;
+
+	  DWORD thread_id = tni->dwThreadID;
+	  if (thread_id == (DWORD)-1)
+	    thread_id = current_event.dwThreadId;
+	  th = thread_rec (thread_id, FALSE);
+
+	  if (th && tni->dwType == 0x1000 && !tni->dwFlags)
+	    {
+	      char *thread_name = NULL;
+	      if (!tni->szName ||
+		  target_read_string((CORE_ADDR)tni->szName, &thread_name,
+				     1024, 0))
+		{
+		  if (thread_name && !*thread_name)
+		    {
+		      xfree (thread_name);
+		      thread_name = NULL;
+		    }
+		  if (th->name)
+		    xfree (th->name);
+		  th->name = thread_name;
+		}
+	    }
+	}
+      return 0;
     default:
       /* Treat unhandled first chance exceptions specially.  */
       if (current_event.u.Exception.dwFirstChance)
@@ -1690,7 +1734,8 @@
 	break;
       if (handle_exception (ourstatus))
 	thread_id = current_event.dwThreadId;
-      else
+      else if (current_event.u.Exception.ExceptionRecord.ExceptionCode
+	       != EXCEPTION_THREAD_NAME)
 	continue_status = DBG_EXCEPTION_NOT_HANDLED;
       break;
 
@@ -2606,6 +2651,13 @@
   return normal_pid_to_str (ptid);
 }
 
+static const char *
+windows_thread_name (struct target_ops *ops, struct thread_info *tp)
+{
+  windows_thread_info *th = thread_rec (ptid_get_tid (tp->ptid), FALSE);
+  return th ? th->name : NULL;
+}
+
 static enum target_xfer_status
 windows_xfer_shared_libraries (struct target_ops *ops,
 			       enum target_object object, const char *annex,
@@ -2712,6 +2764,7 @@
   t->to_mourn_inferior = windows_mourn_inferior;
   t->to_thread_alive = windows_thread_alive;
   t->to_pid_to_str = windows_pid_to_str;
+  t->to_thread_name = windows_thread_name;
   t->to_interrupt = windows_interrupt;
   t->to_pid_to_exec_file = windows_pid_to_exec_file;
   t->to_get_ada_task_ptid = windows_get_ada_task_ptid;
--- a/gdb/gdbserver/win32-low.h	2016-02-10 04:19:39.000000000 +0100
+++ b/gdb/gdbserver/win32-low.h	2016-05-18 18:21:29.420682800 +0200
@@ -37,6 +37,9 @@
   /* Thread Information Block address.  */
   CORE_ADDR thread_local_base;
 
+  /* Thread name.  */
+  char *name;
+
   /* Non zero if SuspendThread was called on this thread.  */
   int suspended;
 
--- a/gdb/gdbserver/win32-low.c	2016-02-10 04:19:39.000000000 +0100
+++ b/gdb/gdbserver/win32-low.c	2016-05-18 19:10:17.365151700 +0200
@@ -236,6 +236,7 @@
 
   remove_thread (thread);
   CloseHandle (th->h);
+  free (th->name);
   free (th);
 }
 
@@ -340,6 +341,62 @@
     return success ? done : -1;
 }
 
+#undef	MIN
+#define MIN(A, B) (((A) <= (B)) ? (A) : (B))
+
+static int
+target_read_string (CORE_ADDR memaddr, char **string, int len)
+{
+  int tlen, offset, i;
+  gdb_byte buf[4];
+  char *buffer;
+  int buffer_allocated;
+  char *bufptr;
+  unsigned int nbytes_read = 0;
+
+  buffer_allocated = 4;
+  buffer = (char *) xmalloc (buffer_allocated);
+  bufptr = buffer;
+
+  while (len > 0)
+    {
+      tlen = MIN (len, 4 - (memaddr & 3));
+      offset = memaddr & 3;
+
+      i = child_xfer_memory (memaddr & ~3, (char *) buf, sizeof buf, 0, 0);
+      if (i != sizeof buf)
+	goto done;
+
+      if (bufptr - buffer + tlen > buffer_allocated)
+	{
+	  unsigned int bytes;
+
+	  bytes = bufptr - buffer;
+	  buffer_allocated *= 2;
+	  buffer = (char *) xrealloc (buffer, buffer_allocated);
+	  bufptr = buffer + bytes;
+	}
+
+      for (i = 0; i < tlen; i++)
+	{
+	  *bufptr++ = buf[i + offset];
+	  if (buf[i + offset] == '\000')
+	    {
+	      nbytes_read += i + 1;
+	      goto done;
+	    }
+	}
+
+      memaddr += tlen;
+      len -= tlen;
+      nbytes_read += tlen;
+    }
+
+done:
+  *string = buffer;
+  return nbytes_read;
+}
+
 /* Clear out any old thread list and reinitialize it to a pristine
    state. */
 static void
@@ -1232,6 +1289,17 @@
   unloaded_dll (NULL, load_addr);
 }
 
+#pragma pack(push,8)
+typedef struct tagTHREADNAME_INFO
+{
+  DWORD dwType;
+  LPCSTR szName;
+  DWORD dwThreadID;
+  DWORD dwFlags;
+} THREADNAME_INFO;
+#pragma pack(pop)
+#define EXCEPTION_THREAD_NAME 0x406d1388
+
 static void
 handle_exception (struct target_waitstatus *ourstatus)
 {
@@ -1322,6 +1390,41 @@
       OUTMSG2 (("EXCEPTION_NONCONTINUABLE_EXCEPTION"));
       ourstatus->value.sig = GDB_SIGNAL_ILL;
       break;
+    case EXCEPTION_THREAD_NAME:
+      if (current_event.u.Exception.ExceptionRecord.NumberParameters
+	  == sizeof (THREADNAME_INFO) / sizeof (ULONG_PTR))
+	{
+	  THREADNAME_INFO *tni = (THREADNAME_INFO *)
+	    current_event.u.Exception.ExceptionRecord.ExceptionInformation;
+
+	  DWORD thread_id = tni->dwThreadID;
+	  ptid_t ptid;
+	  win32_thread_info *th;
+	  if (thread_id == (DWORD)-1)
+	    thread_id = current_event.dwThreadId;
+	  ptid = ptid_build (current_event.dwProcessId, thread_id, 0);
+	  th = thread_rec (ptid, FALSE);
+
+	  if (th && tni->dwType == 0x1000 && !tni->dwFlags)
+	    {
+	      char *thread_name = NULL;
+	      if (!tni->szName ||
+		  target_read_string((size_t)tni->szName, &thread_name,
+				     1024))
+		{
+		  if (thread_name && !*thread_name)
+		    {
+		      xfree (thread_name);
+		      thread_name = NULL;
+		    }
+		  if (th->name)
+		    xfree (th->name);
+		  th->name = thread_name;
+		}
+	    }
+	}
+      ourstatus->kind = TARGET_WAITKIND_SPURIOUS;
+      return;
     default:
       if (current_event.u.Exception.dwFirstChance)
 	{
@@ -1792,6 +1895,13 @@
   return the_low_target.breakpoint;
 }
 
+static const char *
+win32_thread_name (ptid_t thread)
+{
+  win32_thread_info *th = thread_rec (thread, 0);
+  return th ? th->name : NULL;
+}
+
 static struct target_ops win32_target_ops = {
   win32_create_inferior,
   NULL,  /* post_create_inferior */
@@ -1870,6 +1980,7 @@
   NULL, /* multifs_readlink */
   NULL, /* breakpoint_kind_from_pc */
   win32_sw_breakpoint_from_kind,
+  win32_thread_name,
 };
 
 /* Initialize the Win32 backend.  */
