From 306eba957448c74281f41ca8b6f95d76b9f72603 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sat, 29 Aug 2015 16:03:07 +0200
Subject: [PATCH 5/9] free argv[]

---
 mingw-w64-crt/crt/crtexe.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/mingw-w64-crt/crt/crtexe.c b/mingw-w64-crt/crt/crtexe.c
index f4ad70cfe..97014aa79 100644
--- a/mingw-w64-crt/crt/crtexe.c
+++ b/mingw-w64-crt/crt/crtexe.c
@@ -48,6 +48,7 @@ extern int __mingw_app_type;
 static int argc;
 extern void __main(void);
 static _TCHAR **argv;
+static _TCHAR **argv_copy;
 static _TCHAR **envp;
 
 static int managedapp;
@@ -255,6 +256,9 @@ __tmainCRTStartup (void)
 	ret = duplicate_ppstrings (argc, &argv);
 	if (ret != 0)
 	  _amsg_exit (8); /* _RT_SPACEARG */
+	argv_copy = (_TCHAR **) malloc (sizeof (_TCHAR *) * argc);
+	if (argv_copy != NULL)
+	  memcpy (argv_copy, argv, sizeof (_TCHAR *) * argc);
 
 	_initterm (__xc_a, __xc_z);
 	__main (); /* C++ initialization. */
@@ -275,6 +279,14 @@ __tmainCRTStartup (void)
     __initenv = envp;
 #endif
     ret = _tmain (argc, argv, envp);
+    if (argv_copy != NULL)
+    {
+      int i;
+      for (i=0; i < argc; i++)
+        free (argv_copy[i]);
+      free (argv_copy);
+      free (argv);
+    }
     if (!managedapp)
       exit (ret);
 
-- 
2.53.0

