From cfab824f86b10ec6127a04b58f2711be749c42b1 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sat, 29 Aug 2015 16:03:07 +0200
Subject: [PATCH 5/8] free argv[]

---
 mingw-w64-crt/crt/crtexe.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/mingw-w64-crt/crt/crtexe.c b/mingw-w64-crt/crt/crtexe.c
index 9aca06e66..0a5248d3a 100644
--- a/mingw-w64-crt/crt/crtexe.c
+++ b/mingw-w64-crt/crt/crtexe.c
@@ -66,6 +66,7 @@ extern int __mingw_app_type;
 static int argc;
 extern void __main(void);
 static _TCHAR **argv;
+static _TCHAR **argv_copy;
 static _TCHAR **envp;
 
 static int argret;
@@ -276,6 +277,8 @@ __tmainCRTStartup (void)
     _fpreset ();
 
     duplicate_ppstrings (argc, &argv);
+    argv_copy = (_TCHAR **) malloc (sizeof (_TCHAR *) * argc);
+    memcpy (argv_copy, argv, sizeof (_TCHAR *) * argc);
     __main (); /* C++ initialization. */
 #ifdef _UNICODE
     __winitenv = envp;
@@ -283,6 +286,13 @@ __tmainCRTStartup (void)
     __initenv = envp;
 #endif
     mainret = _tmain (argc, argv, envp);
+    {
+      int i;
+      for (i=0; i < argc; i++)
+        free (argv_copy[i]);
+      free (argv_copy);
+      free (argv);
+    }
     if (!managedapp)
       exit (mainret);
 
-- 
2.39.1.windows.1

