From 3b8e6ba7fa90a65b168a168d3357b849ea82f2d1 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sat, 29 Aug 2015 15:54:43 +0200
Subject: [PATCH 06/10] use _TCHAR for argv

---
 mingw-w64-crt/crt/crtexe.c | 52 +++++++---------------------------------------
 1 file changed, 8 insertions(+), 44 deletions(-)

diff --git a/mingw-w64-crt/crt/crtexe.c b/mingw-w64-crt/crt/crtexe.c
index 513ebefb..cc38e3d4 100644
--- a/mingw-w64-crt/crt/crtexe.c
+++ b/mingw-w64-crt/crt/crtexe.c
@@ -88,13 +88,8 @@ DWORD __mingw_winmain_nShowCmd = SW_SHOWDEFAULT;
 
 static int argc;
 extern void __main(void);
-#ifdef WPRFLAG
-static wchar_t **argv;
-static wchar_t **envp;
-#else
-static char **argv;
-static char **envp;
-#endif
+static _TCHAR **argv;
+static _TCHAR **envp;
 
 static int argret;
 static int mainret=0;
@@ -105,11 +100,7 @@ extern LPTOP_LEVEL_EXCEPTION_FILTER __mingw_oldexcpt_handler;
 
 extern void _pei386_runtime_relocator (void);
 long CALLBACK _gnu_exception_handler (EXCEPTION_POINTERS * exception_data);
-#ifdef WPRFLAG
-static void duplicate_ppstrings (int ac, wchar_t ***av);
-#else
-static void duplicate_ppstrings (int ac, char ***av);
-#endif
+static void duplicate_ppstrings (int ac, _TCHAR ***av);
 
 static int __cdecl pre_c_init (void);
 static void __cdecl pre_cpp_init (void);
@@ -408,49 +399,22 @@ check_managed_app (void)
   return 0;
 }
 
-#ifdef WPRFLAG
-static size_t wbytelen(const wchar_t *p)
-{
-	size_t ret = 1;
-	while (*p!=0) {
-		ret++,++p;
-	}
-	return ret*2;
-}
-static void duplicate_ppstrings (int ac, wchar_t ***av)
+static void duplicate_ppstrings (int ac, _TCHAR ***av)
 {
-	wchar_t **avl;
+	_TCHAR **avl;
 	int i;
-	wchar_t **n = (wchar_t **) malloc (sizeof (wchar_t *) * (ac + 1));
+	_TCHAR **n = (_TCHAR **) malloc (sizeof (_TCHAR *) * (ac + 1));
 
 	avl=*av;
 	for (i=0; i < ac; i++)
 	  {
-		size_t l = wbytelen (avl[i]);
-		n[i] = (wchar_t *) malloc (l);
+		size_t l = (_tcslen (avl[i]) + 1) * sizeof (_TCHAR);
+		n[i] = (_TCHAR *) malloc (l);
 		memcpy (n[i], avl[i], l);
 	  }
 	n[i] = NULL;
 	*av = n;
 }
-#else
-static void duplicate_ppstrings (int ac, char ***av)
-{
-	char **avl;
-	int i;
-	char **n = (char **) malloc (sizeof (char *) * (ac + 1));
-	
-	avl=*av;
-	for (i=0; i < ac; i++)
-	  {
-		size_t l = strlen (avl[i]) + 1;
-		n[i] = (char *) malloc (l);
-		memcpy (n[i], avl[i], l);
-	  }
-	n[i] = NULL;
-	*av = n;
-}
-#endif
 
 int __cdecl atexit (_PVFV func)
 {
-- 
2.15.1.windows.2

