From 618f2904bccf3555244500a03b3e3760ea9a7cdb Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Sat, 29 Aug 2015 16:03:07 +0200
Subject: [PATCH 6/9] free argv[]

---
 mingw-w64-crt/crt/crtexe.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/mingw-w64-crt/crt/crtexe.c b/mingw-w64-crt/crt/crtexe.c
index 9cef4e1fb..3abc8ea52 100644
--- a/mingw-w64-crt/crt/crtexe.c
+++ b/mingw-w64-crt/crt/crtexe.c
@@ -71,6 +71,7 @@ extern int __mingw_app_type;
 static int argc;
 extern void __main(void);
 static _TCHAR **argv;
+static _TCHAR **argv_copy;
 static _TCHAR **envp;
 
 static int argret;
@@ -281,6 +282,8 @@ __tmainCRTStartup (void)
     _fpreset ();
 
     duplicate_ppstrings (argc, &argv);
+    argv_copy = (_TCHAR **) malloc (sizeof (_TCHAR *) * argc);
+    memcpy (argv_copy, argv, sizeof (_TCHAR *) * argc);
     __main (); /* C++ initialization. */
 #ifdef _UNICODE
     __winitenv = envp;
@@ -288,6 +291,13 @@ __tmainCRTStartup (void)
     __initenv = envp;
 #endif
     mainret = _tmain (argc, argv, envp);
+    {
+      int i;
+      for (i=0; i < argc; i++)
+        free (argv_copy[i]);
+      free (argv_copy);
+      free (argv);
+    }
     if (!managedapp)
       exit (mainret);
 
-- 
2.39.1.windows.1

