From a5a6905f7f6a65dbf5928d36fa3fb77dfca462f6 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Mon, 21 Mar 2016 19:23:39 +0100
Subject: [PATCH 08/11] fix alignment of SETJMP_FLOAT128

---
 mingw-w64-headers/crt/setjmp.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mingw-w64-headers/crt/setjmp.h b/mingw-w64-headers/crt/setjmp.h
index fc79f109..f9bbcb7d 100644
--- a/mingw-w64-headers/crt/setjmp.h
+++ b/mingw-w64-headers/crt/setjmp.h
@@ -109,7 +109,7 @@ extern "C" {
 
 #elif defined(__x86_64)
 
-  typedef _CRT_ALIGN(16) struct _SETJMP_FLOAT128 {
+  typedef struct _CRT_ALIGN(16) _SETJMP_FLOAT128 {
     __MINGW_EXTENSION unsigned __int64 Part[2];
   } SETJMP_FLOAT128;
 
-- 
2.15.1.windows.2

