From 5725c65d04460de7da27930d1ca3c37c97f539c2 Mon Sep 17 00:00:00 2001
From: Hannes Domani <ssbssa@yahoo.de>
Date: Wed, 23 Mar 2016 17:45:22 +0100
Subject: [PATCH 09/11] dwmapi: add missing *Iconic* functions

---
 mingw-w64-crt/lib64/dwmapi.def | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/mingw-w64-crt/lib64/dwmapi.def b/mingw-w64-crt/lib64/dwmapi.def
index 871947e..6c4059d 100644
--- a/mingw-w64-crt/lib64/dwmapi.def
+++ b/mingw-w64-crt/lib64/dwmapi.def
@@ -17,11 +17,14 @@ DwmGetGraphicsStreamClient
 DwmGetGraphicsStreamTransformHint
 DwmGetTransportAttributes
 DwmGetWindowAttribute
+DwmInvalidateIconicBitmaps
 DwmIsCompositionEnabled
 DwmModifyPreviousDxFrameDuration
 DwmQueryThumbnailSourceSize
 DwmRegisterThumbnail
 DwmSetDxFrameDuration
+DwmSetIconicLivePreviewBitmap
+DwmSetIconicThumbnail
 DwmSetPresentParameters
 DwmSetWindowAttribute
 DwmUnregisterThumbnail
-- 
1.9.4.msysgit.2

